File: 3164-find-the-number-of-good-pairs-ii/3164-find-the-number-of-good-pairs-ii.py
class Solution:
    def numberOfPairs(self, a: List[int], b: List[int], k: int) -> int:
        cnt = [0] * int(1e6+1) 
        for x in b: cnt[x] += 1 
            
        ans = 0 
        for x, c in Counter(a).items():
            for d in range(1,x+1):
                if d * d > x : break 
                
                if x % d == 0:
                    y, z = d, x // d
                    if y % k == 0: ans += cnt[y//k]*c 
                    if z % k == 0 and z != y: ans += cnt[z//k]*c
        return ans

File: 3164-find-the-number-of-good-pairs-ii/3164-find-the-number-of-good-pairs-ii.cpp
using ll = long long;
const int sz = 1e6;
class Solution {
public:
    long long numberOfPairs(vector<int>& a, vector<int>& b, int k) {
        vector<ll> cnt(sz+1);        
        for (auto x : a) {
            for (int d=1; d*d <= x; d++) {
                if (x % d == 0) {
                    int y = d, z = x / d;
                    cnt[y]++;
                    if (y!=z) cnt[z]++;
                }
            }
        }
        ll ans = 0;
        for (auto y : b) {
            if (k*y <= sz) ans += cnt[k*y];
        }
        return ans;        
    }
};